### ISSUE TITLE
OCI_Execute() raises error OCI-22060 when binding host big_int variables with OCI_BDM_OUT in case the SQL or PL/SQL engine sets a NULL value

### ISSUE THREAD

**Issue by Andreas741** [vrogier/ocilib#308]
Hello,

I think I found a bug in obilib 4.7.4.
This error does not occur with version 4.5.1 of ocilib.
Attached is the SQL script and the C++ code.

Kind regards

OcilibTest1.sql

```sql
create table TestTable (nr number, kg_id number);

CREATE OR REPLACE PACKAGE TestPackage IS
    type mf_verlauf_rec is record
    (
        nr number,
        kg_id number
    );
    FUNCTION GetCurrentItem( nr_ number, item_ OUT TestPackage.mf_verlauf_rec) return number;
END;
/

CREATE OR REPLACE PACKAGE BODY TestPackage IS
    function GetCurrentItem( nr_ number, item_ OUT TestPackage.mf_verlauf_rec) return number
    is
        dataset TestTable%rowtype;
    begin
        item_.nr := dataset.NR;
        item_.kg_id := dataset.KG_ID;

        return item_.nr;
    end;
END;
/
```

// OcilibTest1.cpp

```c++
/*
Error occurred at OcilibNumberTranslateValue: OCI-22060: Argument [2] ist eine ung³ltige oder nicht initialisierte Zahl
*/

#include "ocilib.hpp"
#include <iostream>
#include <cstdint>

int main(int argc, char** argv)
{
    if (argc < 4)
        return EXIT_FAILURE;

    const otext *usr = argv[1];
    const otext *pwd = argv[2];
    const otext *db = argv[3];

    try
    {
        ocilib::Environment::Initialize();
        ocilib::Connection connection(db, usr, pwd);
        ocilib::Statement statement(connection);

        int64_t verlauf_nr = 9453;
        int64_t kg_id = 0;  // 1743482
        int64_t ret = 0;
        const ocilib::ostring cmd =
            R"(
declare
    iItem TestPackage.mf_verlauf_rec;
begin
    :ret := TestPackage.GetCurrentItem(:verlauf_nr, iItem);
    :kg_id := iItem.KG_ID;
end;
)";

        statement.Prepare(cmd);
        statement.Bind("verlauf_nr", verlauf_nr, ocilib::BindInfo::In);
        statement.Bind("kg_id", kg_id, ocilib::BindInfo::Out);
        statement.Bind("ret", ret, ocilib::BindInfo::Out);
        statement.ExecutePrepared();
        int x = 0;
    }
    catch (ocilib::Exception e)
    {
        std::cout << e.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
```

**Comment by vrogier** [comment]
Hi,

It seems indeed to be a regression.
I will investigate this asap.

Best regards,

Vincent

**Comment by Andreas741** [comment]
Hello,
many thanks for the quick response

The error message is identical to that from issue #86. 
But there must be another reason. 
I tested the scripts from issue #86 and they work correctly.

Greetings from Germany

**Comment by Andreas741** [comment]
Hello,

i investigated the error further and found the following. 
The error only occurs if the variable kg_id in the C++ code has the type int64_t or uint64_t, 
no error occurs with int32_t or double. 
The error is triggered by the OCI function OCINumberToInt, file number.c in line 162.
As far as I can see the parameters are correct.

Kind regards

**Comment by vrogier** [comment]
Hi,

Do have the issue if you declare kg_id  as ocilib::BindInfo::InOut ?

Regards,

vincent

**Comment by vrogier** [comment]
I tried supplied code (had to modify the package function body which is invalid to assign random values to the record members).
It runs fine on my windows box.

**Comment by vrogier** [comment]
Can you provide a valid PL/SQL definition (as I had modified it an possibly changed the execution context) ?
What compiler and platform are you using ?

Regards,

Vincent

**Comment by Andreas741** [comment]
Hello,

I am using Visual Studio 2019 and 2022 on Windows 10 with Oracle 12.
The error only occurs if the record in the package is not initialized.

I made the following change, so it works now.

file: statement.c
function: static boolean OcilibStatementBindUpdate

        else if (OCI_NUM_BIGINT & bnd->subtype)
        {
            OCINumber *src_number = OCI_BIND_GET_BUFFER(src, OCINumber, index);
            big_int   *dst_bint   = OCI_BIND_GET_SCALAR(dst, big_int, index);
            // akh 01.06.2022
            // Don't call OcilibNumberTranslateValue with null values
            const OCIInd *ind_ptr = bnd->buffer.inds;
            const int is_null = (ind_ptr ? *ind_ptr == OCI_IND_NULL : 0);

            if (dst_bint)
            {
                if (is_null)
                    memset(dst_bint, 0, sizeof(big_int));
                else
                    CHECK(OcilibNumberTranslateValue(bnd->stmt->con, src_number,
                                                 OCI_NUM_NUMBER, dst_bint, bnd->subtype))
            }
        }

Kind regards

Andreas741

**Comment by vrogier** [comment]
Hi,

Do you have a valid PL/SQL declaration and C++ code  I can use to test ?

In the meanwhile, I will apply some change apply bind value change only if it is not null.

Thanks :)

Vincent

**Comment by Andreas741** [comment]
Hello,

I use the code from my 1st post.
I can build and run the package with no errors.
What's wrong with the package?

Kind regards

Andrew741

**Comment by vrogier** [comment]
Hi,

I had to modify the syntax but succeed in the end to reproduce the issue.

I committed a fix and test case for it in v4.7.5 branch

Thanks you !

Regards,

Vincent
